File: src/main/java/com/yuukaze/i18next/ui/dialog/EditDialog.java
package com.yuukaze.i18next.ui.dialog;

import com.intellij.openapi.project.Project;
import com.intellij.openapi.ui.DialogWrapper;
import com.yuukaze.i18next.model.KeyedTranslation;
import com.yuukaze.i18next.model.TranslationDelete;
import com.yuukaze.i18next.model.TranslationUpdate;
import com.yuukaze.i18next.service.DataStore;
import com.yuukaze.i18next.ui.components.LocaleDialogBase;
import com.yuukaze.i18next.ui.dialog.descriptor.DeleteActionDescriptor;
import org.jetbrains.annotations.Nullable;

import java.util.HashMap;
import java.util.Map;
import java.util.ResourceBundle;

/**
 * Edit translation dialog.
 * @author marhali
 */
public class EditDialog extends LocaleDialogBase {
    private final KeyedTranslation origin;

    public EditDialog(Project project, KeyedTranslation origin) {
        super(project,ResourceBundle.getBundle("messages").getString("action.edit"));
        this.origin = origin;
    }

    public void showAndHandle() {
        int code = prepare().show();

        if(code == DialogWrapper.OK_EXIT_CODE) { // Edit
            DataStore.getInstance(project).processUpdate(new TranslationUpdate(origin, getChanges()));

        } else if(code == DeleteActionDescriptor.EXIT_CODE) { // Delete
            DataStore.getInstance(project).processUpdate(new TranslationDelete(origin));
        }
    }

    private KeyedTranslation getChanges() {
        Map<String, String> messages = new HashMap<>();

        valueTextFields.forEach((k, v) -> {
            if(!v.getText().isEmpty()) {
                messages.put(k, v.getText());
            }
        });

        return new KeyedTranslation(keyTextField.getText(), messages);
    }

    @Nullable
    @Override
    protected String getPreKey() {
        return this.origin.getKey();
    }

    @Nullable
    @Override
    protected String getTranslation(String locale) {
        return this.origin.getTranslations().get(locale);
    }
}

File: src/main/kotlin/com/yuukaze/i18next/ui/dialog/EditDialog.kt
package com.yuukaze.i18next.ui.dialog

import com.intellij.openapi.project.Project
import com.intellij.openapi.ui.DialogWrapper
import com.yuukaze.i18next.model.KeyedTranslation
import com.yuukaze.i18next.model.TranslationDelete
import com.yuukaze.i18next.model.TranslationUpdate
import com.yuukaze.i18next.service.DataStore
import com.yuukaze.i18next.ui.components.LocaleDialogBase
import com.yuukaze.i18next.ui.dialog.descriptor.DeleteActionDescriptor
import java.util.*

class EditDialog(project: Project?, private val origin: KeyedTranslation) :
  LocaleDialogBase(
    project!!, ResourceBundle.getBundle("messages").getString("action.edit")
  ) {
  fun showAndHandle() {
    val code = prepare().show()
    if (code == DialogWrapper.OK_EXIT_CODE) { // Edit
      DataStore.getInstance(project).processUpdate(
        TranslationUpdate(
          origin, changes
        )
      )
    } else if (code == DeleteActionDescriptor.EXIT_CODE) { // Delete
      DataStore.getInstance(project).processUpdate(
        TranslationDelete(
          origin
        )
      )
    }
  }

  private val changes: KeyedTranslation
    get() {
      return KeyedTranslation(keyTextField!!.text, entry)
    }
  override val preKey: String?
    get() = origin.key

  override fun getTranslation(locale: String?): String? {
    return origin.translations[locale]
  }
}